File: client/database/answer.model.ts
import { Schema, model, models, Document, Model, Types } from 'mongoose'
import { IUser } from './user.model'
import { IQuestion } from './question.model'

export interface IAnswer extends Document {
  author: Types.ObjectId | IUser
  question: Types.ObjectId | IQuestion
  content: string
  upvotes: Types.ObjectId[]
  downvotes: Types.ObjectId[]
  createdAt: Date
}

const AnswerSchema = new Schema({
  author: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  question: {
    type: Schema.Types.ObjectId,
    ref: 'Question',
    required: true,
  },
  content: {
    type: String,
    required: true,
  },
  upvotes: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  downvotes: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  createdAt: {
    type: Date,
    default: Date.now,
  },
})

export const Answer: Model<IAnswer> = models.Answer || model<IAnswer>('Answer', AnswerSchema)

File: api/internal/data/Answer.model.go
package data

import "time"

type Answer struct {
	Id         int64     `json:"_id"`
	Author     User      `json:"author"`
	AuthorId   int64     `json:"authorId"`
	Question   Question  `json:"question"`
	QuestionId int64     `json:"questionId"`
	Content    string    `json:"content"`
	CreatedAt  time.Time `json:"createdAt"`
	Upvotes    []int64   `json:"upvotes"`
	Downvotes  []int64   `json:"downvotes"`
}

type CreateAnswerParams struct {
	Id         int64
	AuthorId   int64
	QuestionId int64
	Content    string
}

type AnswerVoteParams struct {
	UserId   int64
	AnswerId int64
}
